File: SearchForARange.py
"""
34. Search for a Range
Given a sorted array of integers, find the starting and ending position of a given target value.
Your algorithm's runtime complexity must be in the order of O(log n).
If the target is not found in the array, return [-1, -1].

For example,
Given [5, 7, 7, 8, 8, 10] and target value 8,
return [3, 4].
"""
# O(logn)
# O(1)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums:
            return [-1, -1]
        lower, upper = 0, len(nums) - 1
        while lower < upper:
            mid = (lower + upper) // 2
            if nums[mid] > target:
                upper = mid - 1
            elif nums[mid] < target:
                lower = mid + 1
            else:
                upper = mid
        if nums[lower] != target:
            return [-1, -1]
        left = lower
        lower, upper = 0, len(nums) - 1
        while lower < upper:
            mid = (lower + upper) // 2
            if nums[mid] > target:
                upper = mid - 1
            elif nums[mid] < target:
                lower = mid + 1
            else:
                lower = mid + 1
        if lower > 0 and nums[lower] != target:
            right = lower - 1
        else:
            right = lower
        return [left, right]


if __name__ == "__main__":
    print(Solution().searchRange([5,7,7,8,8,10], 8))

File: Java/SearchForARange.java

public class SearchForARange {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[]{-1, -1};
        if(nums == null || nums.length == 0) {return res;}
        int lower = 0;
        int upper = nums.length - 1;
        int left = binarySearch(lower, upper, nums, target, true);
        if(nums[left] != target) {return res;}
        res[0] = left;
        int right = binarySearch(lower, upper, nums, target, false);
        if(nums[right] == target) {
            res[1] = right;
        } else {
            res[1] = right - 1;
        }
        return res;
    }
    
    private int binarySearch(int lower, int upper, int[] nums, int target, boolean isLeft) {
        int mid;
        while(lower < upper) {
            mid = lower + (upper - lower) / 2;
            if(nums[mid] < target) {
                lower = mid + 1;
            } else if(nums[mid] > target){
                upper = mid - 1;
            } else {
                if(isLeft) {
                    upper = mid;
                } else {
                    lower = mid + 1;
                }
            }
        }
        return lower;
    }

}
